Node iterator

blakebyrnes opened the pull request:
This PR adds two major features:
1) Ability to iterate awaited arrays and iterables
2) Ability to await any element itself and "attach" the result to the underlying node ids

calebjclark commented on files/2-finalized/awaited-base/AwaitedIterator.ts:
```diff
@@ -0,0 +1,47 @@
+    const awaitedPath = state.awaitedPath as AwaitedPath;
+    const awaitedOptions = state.awaitedOptions;
+    const createChild = AwaitedIterator.creators[this.createItemName];
+    return nodeIds.map(nodeId => {
+      const childPath = awaitedPath.withNodeId(nodeId);
+      return createChild(childPath, awaitedOptions) as T;
+    });
```
Why create ALL the child classes that "might" be iterated through. Wouldn't it be much cleaner to create an actual iterator that creates the children as the list is iterated through. And if the for loop is exited then there is no need to finish creating the remainder of children.

calebjclark replied:
It seems to me that this pull request is not creating an "Awaited Iterator". It's really just implementing the concept of attachedId(s) that are generated when the object is "awaited".

NodeList's iterator implementation is distinct from "awaited", well, kinda... obviously it would use the attachedIds from state to iterate on... but if someone tries to iterate on a non-awaited NodeList class (i.e., non-attached) then the iterator could still work, it would just iterate on an empty attachedIds list... or perhaps we should throw an error inside the iterator if the object hasn't been awaited/attached.

blakebyrnes replied:
I started with a true iterator, and it just wouldn't trigger, so I tried to just create an array of elements instead. This approach worked, and I reasoned that 1) converting the array of numbers to wrapper objects was not very expensive, and 2) we would be dealing 99% of the time with pretty small lists, and 3) we could always improve efficiency if we needed to sometime in the future.

Regarding awaited iterator - are you commenting on confusion/unhappiness with the internally used name? If so, it's definitely evolved a ton since we talked about attaching ids, for sure. It is mostly just abstracting the primary iteration and creation of sub-items out of the base classes.

Your last point about simply iterating on an empty list was what I proposed on the phone and I (maybe mistakenly) thought you said "yes, exactly". An error could work too though. Since you can just kind of keep awaiting it and getting something different, I'm not sure which way is clearer

calebjclark commented on src/TsBodyPrinter.ts:
```diff
@@ -99,6 +105,9 @@ export default class TsBodyPrinter {
     this.printConstructor();
     this.printProperties();
     this.printMethods();
+    if (this.buildType === BuildType.base) {
+      this.printPromiseLike();
+    }
```
This seems a bit broad. For example, what happens when DOMParser (https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) is awaited?

Should we limit this to only Node and it's descendants?

blakebyrnes replied:
That's how I did it originally. After our conversation when you said "everything" should be await-able, I took that literally... Node and it's descendants had a lot that didn't get covered, so I would just need help going through all the things that don't extend Node that still need awaited.

blakebyrnes commented:
A few changes:
- Implemented proper iterators
- Activated parentNode children and made document/element IHtmlCollections all abstract so they create objects first and implement the awaited option
- Changed the awaitedIterator and nodeAttachers so they don't get "create" names passed in
- Constructors now specify "constructorNames" in the state. I think we should find a better way to do this in general, but I am not sure what to do quite yet (and probably not part of this PR)
- HtmlCollections cause issues.. they inherit from a dom base class (HtmlCollectionBase) and that causes havoc in the structure and naming since we try to create our own HtmlCollectionBase. I have temporarily specifically called out HtmlCollections in iterators. I need your advice (and probably a future pr) to properly handle base classes

Issues I noticed:
- customizer does not support chrome to change the abstract/on (it has to do with the target not having an originalTarget associated
- Customizer is losing state for things like "abstract" between loads. I have to re-check querySelectorAll every time.. I didn't dive in to figure out why yet

calebjclark commented:
This seems ready to merge... unless we want to quick.y fix the HTMLCollectionBase conflict.

look in files/0-imported/webidl-replacements/HTMLCollection.webidl and HTMLFormControlsCollection.webidl. 

This was an override file I created to split HTMLCollection into HTMLCollection and HTMLCollectionBase. I did this because the W3C spec has HTMLFormControlsCollection inheriting from HTMLCollection then overriding the namedItem type. Typescript barfs at an child class overriding a parent type (I couldn't get it to work).

My hack was to create HTMLCollectionBase that had everything except namedItem, then allow HTMLCollection and HTMLFormControlsCollection to inherit from HTMLCollectionBase.

To solve your problem, just rename HTMLCollectionBase to something else (i.e., `HTMLCollectionWithoutNamedItem`).

blakebyrnes commented:
I fixed the conflict by just renaming the internal base classes to BaseClass. Is that ok for now?
